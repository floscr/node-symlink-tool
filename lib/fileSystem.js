import yaml from 'js-yaml'
import fs from 'fs'
import expandHomeDir from 'expand-home-dir'
import R from 'ramda'
import path from 'path'
import { logErr, logSuccess } from './log'

function loadSettingsFileSync (file) {
  try {
    const expandedFile = expandHomeDir(file)
    const fileContents = fs.readFileSync(expandedFile)
    const yamlObj = yaml.safeLoad(fileContents)
    return yamlObj
  } catch (e) {
    console.log(e)
    process.exit()
  }
}

export function flattenSettingsArray (settings) {
  let files = []
  for (let prop in settings) {
    if (settings.hasOwnProperty(prop) && settings[prop]) {
      files.push(
        ...settings[prop].map(({ src, dst }) => ({ src, dst, type: prop }))
      )
    }
  }
  return files
}

export const loadSettings = R.pipe(
  loadSettingsFileSync,
  flattenSettingsArray
)

export function parentDirExists (file) {
  const parentDir = path.dirname(file)
  return fs.existsSync(parentDir)
}

export async function createLink (file, options) {
  const src = file.src
  const dst = expandHomeDir(file.dst)

  if (!parentDirExists(dst)) {
    logErr('PARENT_DIR_DOES_NOT_EXIST', dst)
    return
  }

  if (!options.silent) {

    logSuccess('File was linked', file.src, file.dst)
  }
}
