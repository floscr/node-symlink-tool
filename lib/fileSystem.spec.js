import R from 'ramda'
import * as fileSystem from './fileSystem'

const yamlSettings = {
  symlink: [
    {
      src: './source',
      dst: '~/temp'
    },
  ],
  hardlink: [
    {
      src: './source',
      dst: '~/temp'
    },
  ]
}

it('should flatten settings to an array of items with the kind mapped', () => {
  const settings = fileSystem.mapKeysTypeProp(yamlSettings)
  expect(settings).toMatchSnapshot()
})
