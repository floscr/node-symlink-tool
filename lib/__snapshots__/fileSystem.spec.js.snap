// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should flatten settings to an array of items with the kind mapped 1`] = `
Array [
  Object {
    "dst": "~/temp",
    "src": "./source",
    "type": "symlink",
  },
  Object {
    "dst": "~/temp",
    "src": "./source",
    "type": "hardlink",
  },
]
`;
